db/active_replicator_common.go

```diff
@@ -24,15 +25,154 @@ type activeReplicatorCommon struct {
 	replicationStats      *BlipSyncStats
 	onReplicatorComplete  ReplicatorCompleteFunc
 	lock                  sync.RWMutex
+	doReconnect           chan struct{}
+	ctx                   context.Context
+	ctxCancel             context.CancelFunc
+}
+
+func newActiveReplicatorCommon(config *ActiveReplicatorConfig, direction ActiveReplicatorDirection) *activeReplicatorCommon {
+
+	var replicationStats *BlipSyncStats
+	if direction == ActiveReplicatorTypePush {
+		replicationStats = BlipSyncStatsForSGRPush(config.ReplicationStatsMap)
+	} else {
+		replicationStats = BlipSyncStatsForSGRPull(config.ReplicationStatsMap)
+	}
+
+	logCtx := context.WithValue(context.Background(), base.LogContextKey{}, base.LogContext{CorrelationID: config.ID + "-" + string(direction)})
+	ctx, ctxCancel := context.WithCancel(logCtx)
+
+	return &activeReplicatorCommon{
+		config:           config,
+		state:            ReplicationStateStopped,
+		doReconnect:      make(chan struct{}),
+		replicationStats: replicationStats,
+		ctx:              ctx,
+		ctxCancel:        ctxCancel,
+	}
+}
+
+func (a *activeReplicatorCommon) _reconnect() {
+	a.doReconnect <- struct{}{}
+}
+
+func (a *activeReplicatorCommon) startReconnector(_startFn func() error) {
+	base.DebugfCtx(a.ctx, base.KeyReplicate, "starting reconnector")
+
+	initialReconnectInterval := time.Second
+	if a.config.InitialReconnectInterval != 0 {
+		initialReconnectInterval = a.config.InitialReconnectInterval
+	}
+	maxReconnectInterval := time.Minute * 5
+	if a.config.MaxReconnectInterval != 0 {
+		maxReconnectInterval = a.config.MaxReconnectInterval
+	}
+
+	ctx := a.ctx
+	var deadlineCancel context.CancelFunc
+	if a.config.TotalReconnectTimeout != 0 {
+		ctx, deadlineCancel = context.WithDeadline(ctx, time.Now().Add(a.config.TotalReconnectTimeout))
+	}
+
+	sleeperFunc := base.SleeperFuncCtx(
+		base.CreateIndefiniteMaxDoublingSleeperFunc(
+			int(initialReconnectInterval.Milliseconds()),
+			int(maxReconnectInterval.Milliseconds())),
+		ctx)
+
+	retryFunc := func() (shouldRetry bool, err error, _ interface{}) {
+		a.lock.Lock()
+		a.replicationStats.NumReconnects.Add(1)
+		err = a._stop()
```

Comment:
Review whether we can minimize the amount of state value changing going on from stop/failed start/reconnecting.